src/audio/eq_iir/iir_hifi3.c

```diff
@@ -0,0 +1,172 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// Copyright(c) 2017 Intel Corporation. All rights reserved.
+//
+// Author: Seppo Ingalsuo <seppo.ingalsuo@linux.intel.com>
+//         Liam Girdwood <liam.r.girdwood@linux.intel.com>
+//         Keyon Jie <yang.jie@linux.intel.com>
+
+#include <stdint.h>
+#include <stddef.h>
+#include <errno.h>
+#include <sof/audio/format.h>
+#include <sof/audio/eq_iir/iir.h>
+#include <user/eq.h>
+
+#if IIR_HIFI3
+
+#include <xtensa/tie/xt_hifi3.h>
+
+/*
+ * Direct form II transposed second order filter block (biquad)
+ *
+ *              +----+                         +---+    +-------+
+ * X(z) ---o--->| b0 |---> + -------------o--->| g |--->| shift |---> Y(z)
+ *         |    +----+     ^              |    +---+    +-------+
+ *         |               |              |
+ *         |            +------+          |
+ *         |            | z^-1 |          |
+ *         |            +------+          |
+ *         |               ^              |
+ *         |    +----+     |     +----+   |
+ *         o--->| b1 |---> + <---| a1 |---o
+ *         |    +----+     ^     +----+   |
+ *         |               |              |
+ *         |            +------+          |
+ *         |            | z^-1 |          |
+ *         |            +------+          |
+ *         |               ^              |
+ *         |    +----+     |     +----+   |
+ *         o--->| b2 |---> + <---| a2 |---+
+ *              +----+           +----+
+ *
+ */
+
+/* Series DF2T IIR */
+
+/* 32 bit data, 32 bit coefficients and 64 bit state variables */
+
+int32_t iir_df2t(struct iir_state_df2t *iir, int32_t x)
+{
+	ae_f64 acc;
+	ae_valign align;
+	ae_f32x2 coef_a2a1;
+	ae_f32x2 coef_b2b1;
+	ae_f32x2 coef_b0shift;
+	ae_f32x2 gain;
+	ae_f32 in;
+	ae_f32 tmp;
+	ae_f32x2 *coefp;
+	ae_f64 *delayp;
+	ae_f32 out = 0;
+	int i;
+	int j;
+	int shift;
+	int nseries = iir->biquads_in_series;
+
+	/* Bypass is set with number of biquads set to zero. */
+	if (!iir->biquads)
+		return x;
+
+	/* Coefficients order in coef[] is {a2, a1, b2, b1, b0, shift, gain} */
+	coefp = (ae_f32x2 *)&iir->coef[0];
+	delayp = (ae_f64 *)&iir->delay[0];
+	in = x;
+	for (j = 0; j < iir->biquads; j += nseries) {
+		for (i = 0; i < nseries; i++) {
+			/* Compute output: Delay is kept Q17.47 while multiply
+			 * instruction gives Q2.30 x Q1.31 -> Q18.46. Need to
+			 * shift delay line values right by one for same align
+			 * as MAC. Store to delay line need to be shifted left
+			 * by one similarly.
+			 */
+			align = AE_LA64_PP(coefp);
+			AE_LA32X2_IP(coef_a2a1, align, coefp);
```

Comment:
I just wonder why we can not use an aligned version?
For which addresses we are afraid of? And why we cannot align them?

Reply:
I realized later that the problem is that we have odd number of int32_t words per biquad. Even if I would be able to ensure the coefficients block is aligned the proceeding from one biquad to next would make next go out of sync. This could be addressed in next ABI major bump. The gain coefficient has the strange Q format that could be fixed as well. 

Some IIR coefficients swapping code would enable moving AE_LA64_PP() out of busy loop but I suspect it would add more overhead than this aligned loads initialization here plus the coefp rewind a bit later.

Reply:
Thanks for clarification